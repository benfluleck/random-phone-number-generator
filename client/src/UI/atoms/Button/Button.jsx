import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import setClassNames from '../../../helpers/utils/setClassNames';

const Button = ({ children, isDisabled, onClick, left, right }) => (
  <Button.Container
    className={setClassNames({
      // eslint-disable-next-line quote-props
      'button': true,
      'is-large': true,
      'is-link': true,
      'is-outlined': true,
      'is-fullwidth': true,
      'is-pulled-left' : left,
      'is-pulled-right' : right,
    })}
    disabled={isDisabled && 'disabled'}
    onClick={onClick}
  >
    {children}
  </Button.Container>
);

Button.Container = styled.button``;

Button.defaultProps = {
  isDisabled: false,
  right: false,
  left: false,
  onClick: () => { },
};

Button.propTypes = {
  children: PropTypes.oneOfType([ PropTypes.string, PropTypes.node ]).isRequired,
  isDisabled: PropTypes.bool,
  onClick: PropTypes.func,
  left: PropTypes.bool,
  right: PropTypes.bool,
};

export default Button;

