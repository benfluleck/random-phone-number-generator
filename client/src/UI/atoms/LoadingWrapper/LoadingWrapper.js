import React from "react";
import ReactLoading from "react-loading";
import styled from 'styled-components';

const LoadingWrapper = () => (
  <ReactLoading type="spin" color="#1a56f0" height='20%' width='20%' className="loading__wrapper" />
);


export default LoadingWrapper;
