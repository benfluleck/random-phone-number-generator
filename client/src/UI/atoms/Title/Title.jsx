import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import setClassNames from '../../../helpers/utils/setClassNames';

const Title = ({ children }) => (
  <Title.Container className={setClassNames({ 'is-size-1': true, 'has-text-centered': true })}>
    {children}
  </Title.Container>
);

export default Title;

Title.Container = styled.h2``;

Title.propTypes = {
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
};
