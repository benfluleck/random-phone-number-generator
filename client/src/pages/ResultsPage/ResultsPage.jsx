import React, { Component, Fragment } from 'react';
import styled from 'styled-components';

import Title from '../../UI/atoms/Title/Title';
import LoadingWrapper from '../../UI/atoms/LoadingWrapper/LoadingWrapper';
import Button from '../../UI/atoms/Button/Button';
import Table from '../../UI/molecules/Table/Table';
import axiosConfig from '../../config/axiosConfig';


class ResultsPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      maximumNumber: '',
      minimumNumber: '',
      numberOfPhonenumbers: '',
      isLoading: true,
    };
    this.handleRedirect = this.handleRedirect.bind(this);
  }


  componentDidMount() {
    axiosConfig
      .phoneNumbers
      .getNumberInfo()
      .then((res) => {
        const { data: { response } } = res;

        this.setState({
          isLoading: false,
          maximumNumber: response.largestNumber,
          minimumNumber: response.smallestNumber,
          numberOfPhonenumbers: response.numberOfPhonenumbers,
        });
      },
      (error) => {
        this.setState({
          isLoading: true,
          error,
        });
        this.props.history.push('/');
      });
  }

  handleDownload(e) {
    e.preventDefault();
    axiosConfig
      .phoneNumbers
      .download()
      .then((response) => {
        const url = window.URL.createObjectURL(new Blob([ response.data ]));
        const link = document.createElement('a');

        link.href = url;
        link.setAttribute('download', 'text.txt');
        document.body.appendChild(link);
        link.click();
      });
  }

  handleRedirect() {
    this.props.history.push('/');
  }


  render() {
    const {
      maximumNumber, minimumNumber, numberOfPhonenumbers, isLoading,
    } = this.state;

    const items = [
      {
        name: 'Smallest Phone Number',
        value: minimumNumber,
      },
      {
        name: 'Largest Phone Number',
        value: maximumNumber,
      },
      {
        name: 'Number of Phone Numbers',
        value: numberOfPhonenumbers,
      },
    ];

    const renderResultsPage = () => (
      <Fragment>
        <ResultsPage.Title>
          <Title>Latest Generated File Info.</Title>
        </ResultsPage.Title>
        <p className=" is-size-4 has-text-grey-light panel-heading">To view all phone numbers, please download the file</p>
        <Table items={items} />
        <ResultsPage.Buttons className="field is-grouped">
          <p className="control">
            <Button onClick={this.handleDownload}>Download File</Button>
          </p>
          <p className="control">
            <Button onClick={this.handleRedirect}>Back To HomePage</Button>
          </p>
        </ResultsPage.Buttons>

      </Fragment>
    );
    return isLoading ? <LoadingWrapper /> : renderResultsPage();

  }
}



ResultsPage.Title = styled.div`
  &::after {
    display: block;
    content: '';
    position: relative;
    height: 1px;
    background: #e4e7ed;
  }
  &:last-child::after {
    display: none;
  }`;

ResultsPage.Buttons = styled.div`
  justify-content: space-between !important;
`

export default ResultsPage;
