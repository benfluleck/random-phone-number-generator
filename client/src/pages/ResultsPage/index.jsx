import React, { Component, Fragment } from 'react';
import styled from 'styled-components';

import Title from '../../UI/atoms/Title/Title';
import LoadingWrapper from '../../UI/atoms/LoadingWrapper/LoadingWrapper'
import Button from '../../UI/atoms/Button/Button'
import Table from '../../UI/molecules/Table/Table'
import axiosConfig from '../../config/axiosConfig';


class ResultsPage extends Component {
  constructor(props) {
    super(props)
    this.state = {
      maximumNumber: '',
      minimumNumber: '',
      numberOfPhonenumbers: '',
      isLoading: true,
    }
  }


  componentDidMount() {
    axiosConfig
      .phoneNumbers
      .getNumberInfo()
      .then((res) => {
        const { data: { response } } = res
        this.setState({
          isLoading: false,
          maximumNumber: response.largestNumber,
          minimumNumber: response.smallestNumber,
          numberOfPhonenumbers: response.numberOfPhonenumbers,
        });
      },
        (error) => {
          this.setState({
            isLoading: true,
            error
          });
        }
      )
  }

  handleDownload(e) {
    e.preventDefault();
    axiosConfig
      .phoneNumbers
      .download()
    .then((response) => {
      const url = window.URL.createObjectURL(new Blob([response.data]));
      const link = document.createElement('a');
      link.href = url;
      link.setAttribute('download', 'text.txt');
      document.body.appendChild(link);
      link.click();
    });

  }




  render() {
    const { maximumNumber, minimumNumber, numberOfPhonenumbers, isLoading } = this.state

    const items = [
      {
        name: 'Smallest Phone Number',
        value: minimumNumber
      },
      {
        name: 'Largest Phone Number',
        value: maximumNumber
      },
      {
        name: 'Number of Phone Numbers',
        value: numberOfPhonenumbers
      },
    ]

    const renderResultsPage = () => (
      <Fragment>
        <ResultsPage.Title>
          <Title>Results</Title>
        </ResultsPage.Title>
        <Table items={items} />
        <div className="field is-grouped">
          <p className="control">
            <Button onClick={this.handleDownload}>Download File</Button>
          </p>
          <p className="control">
            <Button>Back To HomePage</Button>
          </p>
        </div>
      </Fragment>
    )

    {
      return isLoading
        ? <LoadingWrapper />
        : renderResultsPage();
    }

  }
}


ResultsPage.Title = styled.div`
  &::after {
    display: block;
    content: '';
    position: relative;
    height: 1px;
    background: #e4e7ed;
  }
  &:last-child::after {
    display: none;
  }`

export default ResultsPage;
