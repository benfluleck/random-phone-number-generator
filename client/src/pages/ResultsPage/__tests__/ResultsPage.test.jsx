import React from 'react';
import { shallow, mount } from 'enzyme';

import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import ResultsPage from '../ResultsPage';
import LoadingWrapper from '../../../UI/atoms/LoadingWrapper/LoadingWrapper';
import Title from '../../../UI/atoms/Title/Title';
import Button from '../../../UI/atoms/Button/Button';


const historyMock = { push: jest.fn() };

const setup = (props) => {
  const wrapper = shallow(
    <ResultsPage {...props} />
  );

  const mountedWrapper = mount(
    <ResultsPage {...props} />
  );

  return {
    wrapper,
    mountedWrapper,
  };
};


describe('Component - Results Page when isLoading is true', () => {
  test('should render', () => {
    const { wrapper } = setup();

    expect(wrapper.exists()).toBeTruthy();
  });
  test('should render a Loading Wrapper when isLoading is true', () => {
    const { wrapper } = setup();

    expect(wrapper.contains(<LoadingWrapper />)).toBeTruthy();
  });
  describe('Component - Results Page when isLoading is false', () => {
    const { wrapper } = setup();
    const resultsWrapper = wrapper.setState({ isLoading: false });


    test('should render Page when isLoading is false', () => {
      expect(resultsWrapper).toMatchSnapshot();
      expect(resultsWrapper.exists()).toBeTruthy();
    });
    test('should render Page with a Title', () => {
      expect(wrapper.contains(<Title>Latest Generated File Info.</Title>)).toBeTruthy();
    });
    test('should render two Buttons', () => {
      expect(wrapper.find(Button).length).toEqual(2);
    });
    test('should render first button with text download', () => {
      expect(wrapper.find(Button).first().props().children).toEqual('Download File');
    });
    test('should ha Redirect', () => {
      const { mountedWrapper } = setup({ history: historyMock });
      const mountedResultWrapper = mountedWrapper.setState({ isLoading: false });

      const handleRedirectSpy = jest.spyOn(
        mountedResultWrapper.instance(),
        'handleRedirect'
      );

      mountedResultWrapper.instance().handleRedirect();
      expect(handleRedirectSpy).toHaveBeenCalledTimes(1);
    });

    test('should call componentDidMount', () => {
      const { mountedWrapper } = setup({ history: historyMock });
      const mountedResultWrapper = mountedWrapper.setState({ isLoading: false });

      const mock = new MockAdapter(axios);
      const componentDidMountSpy = jest.spyOn(
        mountedResultWrapper.instance(),
        'componentDidMount'
      );

      mountedResultWrapper.instance().componentDidMount();

      mock.onGet().replyOnce(
        200,
        { res: { response: { data: '', link: '' } } }
      );
      expect(componentDidMountSpy).toHaveBeenCalledTimes(1);
    });

    test('should call handle Download action', () => {
      const { mountedWrapper } = setup({ history: historyMock });

      const mountedResultWrapper = mountedWrapper.setState({ isLoading: false });

      const mockEvent = { preventDefault: jest.fn() };

      mountedResultWrapper.find('button').first().simulate('click', mockEvent);

      expect(mountedResultWrapper.text()).toEqual('Latest Generated File Info.');
    });
  });
});
