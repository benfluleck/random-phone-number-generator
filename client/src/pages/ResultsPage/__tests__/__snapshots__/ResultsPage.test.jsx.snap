// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Results Page when isLoading is true Component - Results Page when isLoading is false should render Page when isLoading is false 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ResultsPage />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <ForwardRef>
          <Title>
            Latest Generated File Info.
          </Title>
        </ForwardRef>,
        <p
          className=" is-size-4 has-text-grey-light panel-heading"
        >
          To view all phone numbers, please download the file
        </p>,
        <Table
          items={
            Array [
              Object {
                "name": "Smallest Phone Number",
                "value": "",
              },
              Object {
                "name": "Largest Phone Number",
                "value": "",
              },
              Object {
                "name": "Number of Phone Numbers",
                "value": "",
              },
            ]
          }
        />,
        <ForwardRef
          className="field is-grouped"
        >
          <p
            className="control"
          >
            <Button
              isDisabled={false}
              left={false}
              onClick={[Function]}
              right={false}
            >
              Download File
            </Button>
          </p>
          <p
            className="control"
          >
            <Button
              isDisabled={false}
              left={false}
              onClick={[Function]}
              right={false}
            >
              Back To HomePage
            </Button>
          </p>
        </ForwardRef>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <Title>
            Latest Generated File Info.
          </Title>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Latest Generated File Info.",
          },
          "ref": null,
          "rendered": "Latest Generated File Info.",
          "type": [Function],
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-htpNat",
            "isStatic": true,
            "rules": Array [
              "
  &::after {
    display: block;
    content: '';
    position: relative;
    height: 1px;
    background: #e4e7ed;
  }
  &:last-child::after {
    display: none;
  }",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-htpNat",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "To view all phone numbers, please download the file",
          "className": " is-size-4 has-text-grey-light panel-heading",
        },
        "ref": null,
        "rendered": "To view all phone numbers, please download the file",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "items": Array [
            Object {
              "name": "Smallest Phone Number",
              "value": "",
            },
            Object {
              "name": "Largest Phone Number",
              "value": "",
            },
            Object {
              "name": "Number of Phone Numbers",
              "value": "",
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <p
              className="control"
            >
              <Button
                isDisabled={false}
                left={false}
                onClick={[Function]}
                right={false}
              >
                Download File
              </Button>
            </p>,
            <p
              className="control"
            >
              <Button
                isDisabled={false}
                left={false}
                onClick={[Function]}
                right={false}
              >
                Back To HomePage
              </Button>
            </p>,
          ],
          "className": "field is-grouped",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Button
                isDisabled={false}
                left={false}
                onClick={[Function]}
                right={false}
              >
                Download File
              </Button>,
              "className": "control",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Download File",
                "isDisabled": false,
                "left": false,
                "onClick": [Function],
                "right": false,
              },
              "ref": null,
              "rendered": "Download File",
              "type": [Function],
            },
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Button
                isDisabled={false}
                left={false}
                onClick={[Function]}
                right={false}
              >
                Back To HomePage
              </Button>,
              "className": "control",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Back To HomePage",
                "isDisabled": false,
                "left": false,
                "onClick": [Function],
                "right": false,
              },
              "ref": null,
              "rendered": "Back To HomePage",
              "type": [Function],
            },
            "type": "p",
          },
        ],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bxivhb",
            "isStatic": true,
            "rules": Array [
              "
  justify-content: space-between !important;
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bxivhb",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ForwardRef>
            <Title>
              Latest Generated File Info.
            </Title>
          </ForwardRef>,
          <p
            className=" is-size-4 has-text-grey-light panel-heading"
          >
            To view all phone numbers, please download the file
          </p>,
          <Table
            items={
              Array [
                Object {
                  "name": "Smallest Phone Number",
                  "value": "",
                },
                Object {
                  "name": "Largest Phone Number",
                  "value": "",
                },
                Object {
                  "name": "Number of Phone Numbers",
                  "value": "",
                },
              ]
            }
          />,
          <ForwardRef
            className="field is-grouped"
          >
            <p
              className="control"
            >
              <Button
                isDisabled={false}
                left={false}
                onClick={[Function]}
                right={false}
              >
                Download File
              </Button>
            </p>
            <p
              className="control"
            >
              <Button
                isDisabled={false}
                left={false}
                onClick={[Function]}
                right={false}
              >
                Back To HomePage
              </Button>
            </p>
          </ForwardRef>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Title>
              Latest Generated File Info.
            </Title>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Latest Generated File Info.",
            },
            "ref": null,
            "rendered": "Latest Generated File Info.",
            "type": [Function],
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-htpNat",
              "isStatic": true,
              "rules": Array [
                "
  &::after {
    display: block;
    content: '';
    position: relative;
    height: 1px;
    background: #e4e7ed;
  }
  &:last-child::after {
    display: none;
  }",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-htpNat",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "To view all phone numbers, please download the file",
            "className": " is-size-4 has-text-grey-light panel-heading",
          },
          "ref": null,
          "rendered": "To view all phone numbers, please download the file",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "items": Array [
              Object {
                "name": "Smallest Phone Number",
                "value": "",
              },
              Object {
                "name": "Largest Phone Number",
                "value": "",
              },
              Object {
                "name": "Number of Phone Numbers",
                "value": "",
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <p
                className="control"
              >
                <Button
                  isDisabled={false}
                  left={false}
                  onClick={[Function]}
                  right={false}
                >
                  Download File
                </Button>
              </p>,
              <p
                className="control"
              >
                <Button
                  isDisabled={false}
                  left={false}
                  onClick={[Function]}
                  right={false}
                >
                  Back To HomePage
                </Button>
              </p>,
            ],
            "className": "field is-grouped",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Button
                  isDisabled={false}
                  left={false}
                  onClick={[Function]}
                  right={false}
                >
                  Download File
                </Button>,
                "className": "control",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": "Download File",
                  "isDisabled": false,
                  "left": false,
                  "onClick": [Function],
                  "right": false,
                },
                "ref": null,
                "rendered": "Download File",
                "type": [Function],
              },
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Button
                  isDisabled={false}
                  left={false}
                  onClick={[Function]}
                  right={false}
                >
                  Back To HomePage
                </Button>,
                "className": "control",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": "Back To HomePage",
                  "isDisabled": false,
                  "left": false,
                  "onClick": [Function],
                  "right": false,
                },
                "ref": null,
                "rendered": "Back To HomePage",
                "type": [Function],
              },
              "type": "p",
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bxivhb",
              "isStatic": true,
              "rules": Array [
                "
  justify-content: space-between !important;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bxivhb",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
